pymeasure/instruments/common_base.py

```diff
@@ -0,0 +1,569 @@
+    :param prefix: String to be prefixed to get dynamically configurable
+                   parameters.
+    """
+
+    def __init__(self, fget=None, fset=None, fdel=None, doc=None, fget_params_list=None,
+                 fset_params_list=None, prefix=""):
+        super().__init__(fget, fset, fdel, doc)
+        self.fget_params_list = () if fget_params_list is None else fget_params_list
+        self.fset_params_list = () if fset_params_list is None else fset_params_list
+        self.name = ""
+        self.prefix = prefix
+
+    def __get__(self, obj, objtype=None):
+        if obj is None:
+            # Property return itself when invoked from a class
+            return self
+        if self.fget is None:
+            raise AttributeError(f"Unreadable attribute {self.name}")
+
+        kwargs = {}
+        for attr in self.fget_params_list:
+            attr_instance_name = self.prefix + "_".join([self.name, attr])
+            if hasattr(obj, attr_instance_name):
+                kwargs[attr] = getattr(obj, attr_instance_name)
+        return self.fget(obj, **kwargs)
+
+    def __set__(self, obj, value):
+        if self.fset is None:
+            raise AttributeError(f"Can't set attribute {self.name}")
+        kwargs = {}
+        for attr in self.fset_params_list:
+            attr_instance_name = self.prefix + "_".join([self.name, attr])
+            if hasattr(obj, attr_instance_name):
+                kwargs[attr] = getattr(obj, attr_instance_name)
+        self.fset(obj, value, **kwargs)
+
+    def __set_name__(self, owner, name):
+        self.name = name
+
+
+class CommonBase:
+    """Base class for instruments and channels.
+
+    This class contains everything needed for pymeasure's property creator
+    :meth:`control` and its derivatives :meth:`measurement` and :meth:`setting`.
+    """
+
+    # Variable holding the list of DynamicProperty parameters that are configurable
+    # by users
+    _fget_params_list = ('get_command',
+                         'values',
+                         'map_values',
+                         'get_process',
+                         'command_process',
+                         'check_get_errors')
+
+    _fset_params_list = ('set_command',
+                         'validator',
+                         'values',
+                         'map_values',
+                         'set_process',
+                         'command_process',
+                         'check_set_errors')
+
+    # Prefix used to store reserved variables
+    __reserved_prefix = "___"
+
+    def __init__(self):
+        self._special_names = self._setup_special_names()
+        # Add children from ChildDescriptors.
+        for item, value in inspect.getmembers(self.__class__):
+            if isinstance(value, ChildDescriptor):
+                if isinstance(value.id, (list, tuple)):
+                    for id, cls in zip(value.id, value.cls):
```

Comment:
To be sure, you also have to check if cls is a sequence. Also, use zip(.., strict=True) to make sure zip yells if the iterator lengths don't match. I know, we do this in `children` already, but defense in depth is good, you don't know what happens in between.

Reply:
Ok. First I thought to move the check logic here only, but I think it is good to raise errors already at class creation (invoking `children`) and not only in the init.

Reply:
I noticed, that we can replace the method "children" with a ChildDescriptor itself. That way we reduce one layer and reduce the checks: The child descriptor should alwyas return the pairs as needed (nevertheless, I'll add strict to zip).

Reply:
I would suggest to leave the static method anyway for symmetry with the style used by control/set and also to make future enhancements easier

Reply:
I could add the Child Descriptor as a class attribute to the Common Base, such that it is similar to a static method in usage.

Reply:
Yes, you could, but here I am confused on what we are trying to achieve. Inventing new animals but I miss the purpose. For example, calling it descriptor would be a little misleading because we do follow the python protocol descriptor. I am at home with COVID so I will have some time to think about this, even though my brain is more obfuscated than usual. Please allow me some time and I will get back to you.

Reply:
We need a container storing the information to create the channels of.
We can call it simply "Child" or Child Creator (although it does not create them itself).

I agree, it is not a Descriptor

Reply:
> We need a container storing the information to create the channels of. We can call it simply "Child" or Child Creator (although it does not create them itself).
> 
> I agree, it is not a Descriptor.
> 
> I wish you a good recovery.

@bmoneke, thanks for your patience. Please let me share my thoughts here:

1. Solution as you proposed it is working, in my opinion
2. I would like to propose that the static method or class to define channels to be called `channel` for clarity. IMO, people will understand the generic term channel referred to an instrument more than what is a child, after all this PR is called "Channel implementation added"
3. My preference is to get a dictionary over a list for the reason I provided earlier.
4. You could still go with the descriptor approach, if in the `__init__`, you don't replace the attribute, but you define an internal one prefixed wtih "_". It looks cleaner to me, but I do not have strong arguments.
5. For dynamic channels management, as discussed earlier, I would consider using the `add_child/remove_child` programmatically
6. Limit the scope of `remove_child` to channels created programmatically by implementers using the `add_child` API, ( not created with channel descriptor approach).

Reply:
>     1. Solution as you proposed it is working, in my opinion

Do you mean the solution, to add the "ChannelCreator" as an attribute to the CommonClass?

>     2. I would like to propose that the static method or class to define channels to be called `channel` for clarity. IMO, people will understand the generic term channel referred to an instrument more than what is a child, after all this PR is called "Channel implementation added"

The name "Channel" is already used for the base class of channels. What about "ChannelCreator"?

>     3. My preference is to get a dictionary over a list for the reason I provided earlier.

In the other discussion I asked, whether exchanging the list for a dictionary would be fine and waited for responses (@bilderbuchi) before implementing.

>     4. You could still go with the descriptor approach, if in the `__init__`, you don't replace the attribute, but you define an internal one prefixed wtih "_". It looks cleaner to me, but I do not have strong arguments.

I don't see the advantage of a descriptor approach in comparison to the "Creator" approach. It makes it more complicated (with the additional names).

>     5. For dynamic channels management, as discussed earlier, I would consider using the `add_child/remove_child` programmatically

That's what is planned (and actually already written in the documentation).

>     6. Limit the scope of `remove_child` to channels created programmatically by implementers using the `add_child` API, ( not created with channel descriptor approach).

Why do you want to limit `remove_child` to those channels?

Reply:
> Do you mean the solution, to add the "ChannelCreator" as an attribute to the CommonClass?
> 

Yes

> 
> The name "Channel" is already used for the base class of channels. What about "ChannelCreator"?

It looks good to me.

>
> 
> Why do you want to limit `remove_child` to those channels?

Because I consider this API reserved for people who knows exactly what they're doing. It handled very specific corner cases and I expect that people using this API will not create channels with `ChannelCreator`. In other words, it's reserved for advanced use cases, while ChannelCreator is the basic and I expect mostly used mode.

Reply:
> I am at home with COVID so I will have some time to think about this, even though my brain is more obfuscated than usual. Please allow me some time and I will get back to you.

Get well soon, and be sure to rest! This thing can drag out a long time (as you surely know).

Reply:
`ChannelCreator` sounds good and clear, agreed!

> Because I consider this API reserved for people who knows exactly what they're doing. It handled very specific corner cases and I expect that people using this API will not create channels with ChannelCreator. In other words, it's reserved for advanced use cases, while ChannelCreator is the basic and I expect mostly used mode.

I agree with this. It could also serve to distinguish between "hardcoded"/builtin channels, and "dynamic" ones.

I think it's worth to do only if it does not complicate the existing implementation too much. This is Python, after all, you can do soo many "unplanned" things to existing objects -- powerful, but often hard to "lock down" reserved API like e.g. in C++.

Reply:
Right now, the channels via ChannelCreator are added via `add_child`, such that there is no difference.
I could add a `protected` attribute to a channel, such that `remove_child` does not remove it.

Reply:
> > I am at home with COVID so I will have some time to think about this, even though my brain is more obfuscated than usual. Please allow me some time and I will get back to you.
> 
> Get well soon, and be sure to rest! This thing can drag out a long time (as you surely know).
Thanks for your message, I missed it between the cracks. I think, I am on my way to recover, finally

Reply:
> Right now, the channels via ChannelCreator are added via `add_child`, such that there is no difference.
> I could add a `protected` attribute to a channel, such that `remove_child` does not remove it.

Sounds good. I think we should do it privately (i.e. `Channel._protected`) to keep this internal/out of the way, and in a hidden way (i.e. when after calling `add_child`, set the attribute on the just created channel instance *directly*), to not enlarge the signature  of `add_channel`. What do you think?

Reply:
Sounds good.
If a channel has that attribute, "remove_child" will raise an error (I'd say a Value Error, as that Child is an invalid value, maybe there are better errors, I have to check).

Thanks for the hint of adding it while reading the Creators, as I already looked (yet unsuccessful) for an option to add it, without changing "add_child".

Reply:
A `TypeError` sounds most appropriate:
> exception TypeError
>    Raised when an operation or function is applied to an object of inappropriate type. The associated value is a string giving details about the type mismatch.
>    This exception may be raised by user code to indicate that an attempted operation on an object is not supported, and is not meant to be. If an object is meant to support a given operation but has not yet provided an implementation, [NotImplementedError](https://devdocs.io/python~3.11/library/exceptions#NotImplementedError) is the proper exception to raise.
>    Passing arguments of the wrong type (e.g. passing a [list](https://devdocs.io/python~3.11/library/stdtypes#list) when an [int](https://devdocs.io/python~3.11/library/functions#int) is expected) should result in a [TypeError](https://devdocs.io/python~3.11/library/exceptions#TypeError), but passing arguments with the wrong value (e.g. a number outside expected boundaries) should result in a [ValueError](https://devdocs.io/python~3.11/library/exceptions#ValueError).

Reply:
But the type (Channel) is appropriate, the "value" (protected channel vs. Free channel) is inappropriate. Therefore a Value Error seems to fit better in our case. 
One could argue, that a protected channel is a different type, though.

Reply:
> One could argue, that a protected channel is a different type, though, such that I could accept a Type Error as well (I prefer a Value Error).

Yeah, that's the angle I am coming from (although I would not go so far as to _really_ make those into two different classes/types. It's not a big deal, though, people will primarily register that an exception happened, and will not look at the exception type. :sweat_smile:

Reply:
Done